engine/access/rpc/backend/connection_factory.go

```diff
@@ -74,24 +88,63 @@ func (cf *ConnectionFactoryImpl) createConnection(address string, timeout time.D
 
 func (cf *ConnectionFactoryImpl) retrieveConnection(grpcAddress string, timeout time.Duration) (*grpc.ClientConn, error) {
 	var conn *grpc.ClientConn
+	var mutex *sync.Mutex
```

Comment:
I think we could simplify this locking logic. there seem to be 2 main options:

1. Optimistic locking (Kan's suggestion) - optimize for infrequent client creation/refreshes
We use a single global lock, but only in the case where we need to add a new client. When adding a new client, take the lock, and check again if a client exists in the cache. If it does, return otherwise add it. When refreshing clients, we'll need to take the global lock. 

2. Per client locking - optimize for more frequent client creation/refreshes
We still need a global lock to ensure the get/add operations are atomic, but then switch to a per client lock. Essentially the flow would be:
* use `PeekOrAdd` to get the client object (or add a new one if it doesn't exist)
* take the client specific lock
* use `Get` to get the client again (this makes sure we update the access metadata)
* check if the connection is healthy. if it is return.
* otherwise setup the connection
* release the client lock

Looking over the grpc data on testnet/mainnet, the vast majority of api calls are for endpoints that use execution nodes, and there are a small number of nodes so they would always be in the cache. Send transaction is a relatively infrequent request, so the collection nodes would be the ones getting evicted.

If we increased the size of the cache so there were less frequent LN evictions, and found a fix for the frequent client invalidations, then I think 1 would be fine.